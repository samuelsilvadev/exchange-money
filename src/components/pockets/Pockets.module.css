.pockets {
	display: flex;
	flex-wrap: wrap;
	justify-content: space-between;
	align-items: flex-start;
	align-content: flex-start;
}

.title {
	margin-bottom: 5rem;
	width: 100%;
}

.pocketSelect {
	width: 10rem;
}

.pocketBallance {
	font-size: 1.6rem;
	text-align: right;
	padding: 0.5rem 0 0.25rem;
}

.pocketBallanceDescription {
	font-size: 1.4rem;
	color: var(--gray);
}
