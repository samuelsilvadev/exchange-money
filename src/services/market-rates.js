export function getMarketRates() {
	// TODO: Update endpoint to come from a API_DEFINITIONS file.
	return fetch(
		`https://openexchangerates.org/api/latest.json?app_id=${process.env.RAZZLE_OPEN_EXCHANGE_RATES_APP_ID}`
	).then((data) => data.json());
}
