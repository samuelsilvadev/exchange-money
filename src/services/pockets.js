export const initialState = [
	{
		label: 'USD',
		value: 1000,
	},
	{
		label: 'GBP',
		value: 5000,
	},
	{
		label: 'EUR',
		value: 2000,
	},
];

export function getPockets() {
	return Promise.resolve(initialState);
}
