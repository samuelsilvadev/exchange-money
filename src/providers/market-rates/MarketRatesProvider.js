import React, {
	createContext,
	useContext,
	useEffect,
	useMemo,
	useState,
} from 'react';

import { getMarketRates } from '../../services/market-rates';

/**
 * The market rates context description.
 *
 * @typedef {Object} MarketRatesContext
 * @property {boolean} isLoading
 * @property {boolean} hasError
 * @property {Object.<string, number>} result
 */

const MarketRatesContext = createContext();

export function MarketRatesProvider(props) {
	const [marketRates, setMarketRates] = useState({});
	const [isLoading, setIsLoading] = useState(false);
	const [hasError, setHasError] = useState(false);

	useEffect(() => {
		setIsLoading(true);
		setHasError(false);

		getMarketRates()
			.then((data) => {
				if (data?.rates) {
					setMarketRates(data.rates);
				}
			})
			.catch(() => setHasError(true))
			.finally(() => setIsLoading(false));
	}, []);

	const context = useMemo(() => {
		return {
			isLoading,
			hasError,
			result: marketRates,
		};
	}, [isLoading, hasError, marketRates]);

	return <MarketRatesContext.Provider value={context} {...props} />;
}

/**
 * Get market rates from global storage.
 *
 * @returns {MarketRatesContext} - The market rates context result.
 */
export function useMarketRatesContext() {
	const marketRates = useContext(MarketRatesContext);

	if (typeof marketRates === 'undefined') {
		throw new TypeError(
			'To use the MarketRatesContext correctly you should wrap your component with <MarketRatesProvider />'
		);
	}

	return marketRates;
}
