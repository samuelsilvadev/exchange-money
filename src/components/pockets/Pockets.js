import React, { useEffect, useState } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

import { usePocketsContext } from '../../providers/pockets/PocketsProvider';
import Select from '../select/Select';

import styles from './Pockets.module.css';

function Pockets(props) {
	const { className } = props;

	const { result: pockets } = usePocketsContext();

	const [selectedPocket, setSelectedPocket] = useState(pockets?.[0]);

	useEffect(() => {
		if (pockets.length > 0) {
			const updatedPocket = pockets.find(
				(pocket) => pocket.label === selectedPocket?.label
			);

			setSelectedPocket(updatedPocket ?? pockets[0]);
		}
	}, [pockets, selectedPocket]);

	const handleChange = (event) => {
		const label = event.target.value;

		const nextSelectedPocket = pockets.find(
			(pocket) => pocket.label === label
		);

		setSelectedPocket(nextSelectedPocket);
	};

	return (
		<section
			className={classnames(styles.pockets, { [className]: className })}
		>
			<h1 className={styles.title}>Pockets</h1>
			<Select
				id="pocket"
				label="Pocket"
				className={styles.pocketSelect}
				onBlur={handleChange}
				onChange={handleChange}
			>
				{pockets.map(({ label }) => {
					return (
						<option value={label} key={label}>
							{label}
						</option>
					);
				})}
			</Select>
			<div>
				<p className={styles.pocketBallance}>
					{selectedPocket?.value ?? '-'}
				</p>
				<p className={styles.pocketBallanceDescription}>
					Total Ballance
				</p>
			</div>
		</section>
	);
}

Pockets.propTypes = {
	className: PropTypes.string,
};

export default Pockets;
