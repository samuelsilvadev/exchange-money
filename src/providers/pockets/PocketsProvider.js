import React, {
	createContext,
	useCallback,
	useContext,
	useEffect,
	useMemo,
	useState,
} from 'react';

import { getPockets } from '../../services/pockets';

/**
 * A pocket description.
 *
 * @typedef {Object} Pocket
 * @property {string} label
 * @property {number} value
 */

/**
 * The pocket context description.
 *
 * @typedef {Object} PocketsContext
 * @property {boolean} isLoading
 * @property {boolean} hasError
 * @property {Array<Pocket>} result
 * @property {Function} transfer
 */

const checkForErrorOnTransferValues = ({ pocket, value }) => {
	if (pocket.value === 0) {
		return new TypeError(`Your ${pocket.label} pocket is empty`);
	}

	if (pocket.value < value) {
		return new TypeError(
			`Insufficient money on your ${pocket.label} pocket`
		);
	}

	return null;
};

const PocketsContext = createContext();

export function PocketsProvider(props) {
	const [pockets, setPockets] = useState([]);
	const [isLoading, setIsLoading] = useState(false);
	const [hasError, setHasError] = useState(false);

	useEffect(() => {
		setIsLoading(true);
		setHasError(false);

		getPockets()
			.then((data) => {
				if (data) {
					setPockets(data);
				}
			})
			.catch(() => setHasError(true))
			.finally(() => setIsLoading(false));
	}, []);

	const transfer = useCallback(
		(from, to) => {
			setIsLoading(true);
			setHasError(false);

			return new Promise((resolve, reject) => {
				const updatedPockets = pockets.reduce(
					(updatedPockets, currentPocket) => {
						const pocket = { ...currentPocket };

						if (pocket.label === from.currency) {
							const parsedFromMoneyAmount = parseFloat(
								from.moneyAmount
							);
							const error = checkForErrorOnTransferValues({
								pocket,
								value: parsedFromMoneyAmount,
							});

							if (error) {
								return reject(error);
							}

							pocket.value -= parsedFromMoneyAmount;
						} else if (pocket.label === to.currency) {
							pocket.value += parseFloat(to.moneyAmount);
						}

						updatedPockets.push(pocket);

						return updatedPockets;
					},
					[]
				);

				setPockets(updatedPockets);

				resolve();
			})
				.catch((error) => {
					setHasError(true);

					return error;
				})
				.finally(() => setIsLoading(false));
		},
		[pockets]
	);

	const context = useMemo(() => {
		return {
			isLoading,
			hasError,
			result: pockets,
			transfer,
		};
	}, [isLoading, hasError, pockets, transfer]);

	return <PocketsContext.Provider value={context} {...props} />;
}

/**
 * Get pocket values from global storage.
 *
 * @returns {PocketsContext} - Result set of pockets objects.
 */
export function usePocketsContext() {
	const pockets = useContext(PocketsContext);

	if (typeof pockets === 'undefined') {
		throw new TypeError(
			'To use the PocketsContext correctly you should wrap your component with <PocketsProvider />'
		);
	}

	return pockets;
}
