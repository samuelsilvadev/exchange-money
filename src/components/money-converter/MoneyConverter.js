import React, { useEffect, useReducer, useState } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

import { useMarketRatesContext } from '../../providers/market-rates/MarketRatesProvider';
import MoneyConverterPanel from '../money-converter-panel/MoneyConverterPanel';
import Swap from '../icons/Swap';

import styles from './MoneyConverter.module.css';
import { usePocketsContext } from '../../providers/pockets/PocketsProvider';
import stringToDecimal from '../../formatter/stringToDecimal';

const ACTION_TYPES = {
	SET_FROM_CURRENCY: 'SET_FROM_CURRENCY',
	SET_TO_CURRENCY: 'SET_TO_CURRENCY',
	SET_FROM_MONEY_AMOUNT: 'SET_FROM_MONEY_AMOUNT',
	SET_TO_MONEY_AMOUNT: 'SET_TO_MONEY_AMOUNT',
	SWAP_CURRENCIES: 'SWAP_CURRENCIES',
	RESET_MONEY_AMOUNT: 'RESET_MONEY_AMOUNT',
};

const INITIAL_STATE = {
	from: {
		currency: '',
		moneyAmount: '',
	},
	to: {
		currency: '',
		moneyAmount: '',
	},
};

function reducer(state, action) {
	switch (action.type) {
		case ACTION_TYPES.SET_FROM_CURRENCY: {
			return {
				...state,
				from: {
					...state.from,
					currency: action.payload,
				},
			};
		}
		case ACTION_TYPES.SET_TO_CURRENCY: {
			return {
				...state,
				to: {
					...state.to,
					currency: action.payload,
				},
			};
		}
		case ACTION_TYPES.SET_FROM_MONEY_AMOUNT: {
			return {
				...state,
				from: {
					...state.from,
					moneyAmount: action.payload,
				},
			};
		}
		case ACTION_TYPES.SET_TO_MONEY_AMOUNT: {
			return {
				...state,
				to: {
					...state.to,
					moneyAmount: action.payload,
				},
			};
		}
		case ACTION_TYPES.SWAP_CURRENCIES: {
			const [nextToCurrency, nextFromCurrency] = [
				state.from.currency,
				state.to.currency,
			];

			return {
				...state,
				from: {
					...state.from,
					currency: nextFromCurrency,
				},
				to: {
					...state.to,
					currency: nextToCurrency,
				},
			};
		}
		case ACTION_TYPES.RESET_MONEY_AMOUNT: {
			return {
				from: {
					...state.from,
					moneyAmount: '',
				},
				to: {
					...state.to,
					moneyAmount: '',
				},
			};
		}
	}
}

const ERROR_MESSAGES = {
	FILL_UP_SPEND_FIELD: 'You need to fill up the spend field',
	CHANGE_CURRENCIES: 'You need to change the one of the currencies',
};

function calculateCurrentRate({ from, to, rates = {}, base = 'USD' }) {
	if (!from.currency || !to.currency) {
		return;
	}

	if (from.currency === base) {
		return rates[to.currency];
	}

	if (to.currency === base) {
		return 1 / rates[from.currency];
	}

	return rates[to.currency] * (1 / rates[from.currency]);
}

function formatCurrentRate(rate) {
	return rate?.toString()?.slice(0, 6);
}

function MoneyConverter(props) {
	const { className } = props;

	const [state, dispatch] = useReducer(reducer, INITIAL_STATE);
	const [submitFormError, setSubmitFormError] = useState();

	const { transfer, isLoading: isPocketsLoading } = usePocketsContext();
	const {
		result: marketRates,
		isLoading: isMarketsRatesLoading,
	} = useMarketRatesContext();
	const currentRate = calculateCurrentRate({
		from: state.from,
		to: state.to,
		rates: marketRates,
	});
	const hasCurrentRate = typeof currentRate !== 'undefined';

	const handleChangeFactory = (type) => (value) => {
		dispatch({ type, payload: value });
	};

	const handleSwapClick = () => {
		dispatch({ type: ACTION_TYPES.SWAP_CURRENCIES });
	};

	const handleOnConvertMoney = (event) => {
		event.preventDefault();

		setSubmitFormError();

		const { to, from } = state;

		if (!to.moneyAmount || !from.moneyAmount) {
			setSubmitFormError(ERROR_MESSAGES.FILL_UP_SPEND_FIELD);
			return;
		}

		if (to.currency === from.currency) {
			setSubmitFormError(ERROR_MESSAGES.CHANGE_CURRENCIES);
			return;
		}

		transfer(from, to)
			.then(() => {
				dispatch({ type: ACTION_TYPES.RESET_MONEY_AMOUNT });
			})
			.catch((error) => {
				setSubmitFormError(error.message);
			});
	};

	useEffect(() => {
		const valueToReceive = state.from.moneyAmount * currentRate;
		const formattedMoneyAmount = state.from.moneyAmount
			? stringToDecimal(valueToReceive.toString())
			: '';

		dispatch({
			type: ACTION_TYPES.SET_TO_MONEY_AMOUNT,
			payload: formattedMoneyAmount,
		});
	}, [state.from.moneyAmount, currentRate]);

	return (
		<form
			onSubmit={handleOnConvertMoney}
			className={classnames(styles.wrapper, { [className]: className })}
		>
			{submitFormError && (
				<p className={styles.errorPanel} role="alert">
					{submitFormError}
				</p>
			)}
			<div className={styles.panelsWrapper}>
				<MoneyConverterPanel
					className={styles.panel}
					label="You spend"
					handlers={{
						currency: handleChangeFactory(
							ACTION_TYPES.SET_FROM_CURRENCY
						),
						moneyAmount: handleChangeFactory(
							ACTION_TYPES.SET_FROM_MONEY_AMOUNT
						),
					}}
					values={{
						currency: state.from.currency,
						moneyAmount: state.from.moneyAmount,
					}}
				/>
				<button
					type="button"
					className={styles.swapButton}
					aria-label="Swap currencies"
					onClick={handleSwapClick}
				>
					<Swap />
				</button>
				<MoneyConverterPanel
					className={styles.panel}
					label="You get"
					handlers={{
						currency: handleChangeFactory(
							ACTION_TYPES.SET_TO_CURRENCY
						),
						moneyAmount: handleChangeFactory(
							ACTION_TYPES.SET_TO_MONEY_AMOUNT
						),
					}}
					values={{
						currency: state.to.currency,
						moneyAmount: state.to.moneyAmount,
					}}
				/>
			</div>
			<div className={styles.marketRateWrapper}>
				<p className={styles.marketRateDescription}>Market Rate</p>
				<p className={styles.marketRateValue}>
					{hasCurrentRate
						? `1 ${state.from.currency} = ${formatCurrentRate(
								currentRate
						  )} ${state.to.currency}`
						: '-'}
				</p>
			</div>
			<button
				disabled={isPocketsLoading || isMarketsRatesLoading}
				className={styles.converterButton}
			>
				Converter Now
			</button>
		</form>
	);
}

MoneyConverter.propTypes = {
	className: PropTypes.string,
};

export default MoneyConverter;
