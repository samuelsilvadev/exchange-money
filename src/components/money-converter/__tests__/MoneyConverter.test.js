import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';

import { MarketRatesProvider } from '../../../providers/market-rates/MarketRatesProvider';
import { PocketsProvider } from '../../../providers/pockets/PocketsProvider';
import MoneyConverter from '../MoneyConverter';

jest.mock('../../../services/market-rates', () => ({
	getMarketRates: () => Promise.resolve({ rates: { USD: 1, EUR: 0.8765 } }),
}));

jest.mock('../../../services/pockets', () => ({
	getPockets: () =>
		Promise.resolve([
			{
				label: 'USD',
				value: 1000,
			},
			{
				label: 'GBP',
				value: 5000,
			},
			{
				label: 'EUR',
				value: 2000,
			},
		]),
}));

describe('<MoneyConverter />', () => {
	it('should render correctly', async () => {
		const {
			container,
			getAllByLabelText,
			getByLabelText,
			getByText,
		} = render(
			<PocketsProvider>
				<MarketRatesProvider>
					<MoneyConverter />
				</MarketRatesProvider>
			</PocketsProvider>
		);

		await waitFor(() => {
			expect(getByText('You spend')).toBeVisible();
			expect(getByText('You get')).toBeVisible();
			expect(getByText('Converter Now')).toBeVisible();
			expect(getByText('Market Rate')).toBeVisible();
			expect(getByLabelText('Swap currencies')).toBeVisible();
			expect(getAllByLabelText('Currency')).toHaveLength(2);
		});

		expect(container.firstChild).toMatchSnapshot();
	});

	it('should show correct market rate values', async () => {
		const { getByText, getAllByLabelText } = render(
			<PocketsProvider>
				<MarketRatesProvider>
					<MoneyConverter />
				</MarketRatesProvider>
			</PocketsProvider>
		);

		let toSelect;

		await waitFor(() => {
			toSelect = getAllByLabelText('Currency')[1];

			expect(getByText('1 USD = 1 USD'));
		});

		fireEvent.change(toSelect, { target: { value: 'EUR' } });

		expect(getByText('1 USD = 0.8765 EUR'));
	});

	it('should display how much I will receive when make the conversion', async () => {
		const { getByLabelText, getAllByLabelText } = render(
			<PocketsProvider>
				<MarketRatesProvider>
					<MoneyConverter />
				</MarketRatesProvider>
			</PocketsProvider>
		);

		let fromInput;
		let toInput;
		let toSelect;

		await waitFor(() => {
			fromInput = getByLabelText('You spend');
			toInput = getByLabelText('You get');
			toSelect = getAllByLabelText('Currency')[1];
		});

		fireEvent.change(toSelect, { target: { value: 'EUR' } });
		fireEvent.change(fromInput, { target: { value: '10' } });

		expect(toInput).toHaveValue(8.76);
	});
});
