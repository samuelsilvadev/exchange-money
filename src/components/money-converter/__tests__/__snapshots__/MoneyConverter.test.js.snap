// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MoneyConverter /> should render correctly 1`] = `
<form
  class="wrapper"
>
  <div
    class="panelsWrapper"
  >
    <div
      class="panel"
    >
      <div
        class="selectWrapper select"
      >
        <label
          class="selectLabel"
          for="converter-panel-You spend"
        >
          Currency
        </label>
        <div
          class="selectIconWrapper"
        >
          <select
            class="select"
            id="converter-panel-You spend"
          >
            <option
              value="USD"
            >
              USD
            </option>
            <option
              value="GBP"
            >
              GBP
            </option>
            <option
              value="EUR"
            >
              EUR
            </option>
          </select>
          <svg
            aria-hidden="true"
            class="selectIcon"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 15l-4.243-4.243 1.415-1.414L12 12.172l2.828-2.829 1.415 1.414z"
            />
          </svg>
        </div>
      </div>
      <div
        class="inputWrapper"
      >
        <label
          class="label"
          for="converter-panel-input-You spend"
        >
          You spend
        </label>
        <input
          class="input"
          id="converter-panel-input-You spend"
          type="number"
          value=""
        />
      </div>
    </div>
    <button
      aria-label="Swap currencies"
      class="swapButton"
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10-4.477 10-10 10zm0-2a8 8 0 1 0 0-16 8 8 0 0 0 0 16zm-5-7h9v2h-4v3l-5-5zm5-4V6l5 5H8V9h4z"
        />
      </svg>
    </button>
    <div
      class="panel"
    >
      <div
        class="selectWrapper select"
      >
        <label
          class="selectLabel"
          for="converter-panel-You get"
        >
          Currency
        </label>
        <div
          class="selectIconWrapper"
        >
          <select
            class="select"
            id="converter-panel-You get"
          >
            <option
              value="USD"
            >
              USD
            </option>
            <option
              value="GBP"
            >
              GBP
            </option>
            <option
              value="EUR"
            >
              EUR
            </option>
          </select>
          <svg
            aria-hidden="true"
            class="selectIcon"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 15l-4.243-4.243 1.415-1.414L12 12.172l2.828-2.829 1.415 1.414z"
            />
          </svg>
        </div>
      </div>
      <div
        class="inputWrapper"
      >
        <label
          class="label"
          for="converter-panel-input-You get"
        >
          You get
        </label>
        <input
          class="input"
          id="converter-panel-input-You get"
          type="number"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    class="marketRateWrapper"
  >
    <p
      class="marketRateDescription"
    >
      Market Rate
    </p>
    <p
      class="marketRateValue"
    >
      1 USD = 1 USD
    </p>
  </div>
  <button
    class="converterButton"
  >
    Converter Now
  </button>
</form>
`;
